src/librustc/ty/sty.rs

```diff
@@ -927,11 +888,15 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
         }
     }
 
-    pub fn is_empty(&self, _cx: TyCtxt) -> bool {
+    pub fn is_empty(&self, cx: TyCtxt) -> bool {
         // FIXME(#24885): be smarter here
         match self.sty {
             TyEnum(def, _) | TyStruct(def, _) => def.is_empty(),
-            _ => false
+            TyEmpty => true,
+            TyTuple(ref tys) => tys.iter().any(|ty| ty.is_empty(cx)),
```

Comment:
> To make it smarter, as the FIXME suggests.

That shouldn't be part of this PR, it introduces certain risks. The code is not prepared to deal with something like this. We need to wait until old trans is gone and then properly introduce enum optimizations around `!` and `Void`.

> Well there's other places where I check for `!` specifically by matching on `sty`.

It should be uniform, otherwise it gets confusing when it's okay to use which.